doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,339 @@ Applying theorems
-      not perform any beta-reduction either.
+.. tacn:: apply {+, {? > } @one_term {? with @bindings } } {? @in_hyp_as }
 
-   .. tacv:: notypeclasses refine @term
-      :name: notypeclasses refine
+   .. insertprodn in_hyp_as as_ipat
 
-      This tactic behaves like :tacn:`refine` except it performs type checking without
-      resolution of typeclasses.
+   .. prodn::
+      in_hyp_as ::= in {+, @ident {? @as_ipat } }
+      as_ipat ::= as @simple_intropattern
+
+   Uses unification to match the type of each :n:`@one_term` with the goal
+   (to do backward reasoning) or with a hypothesis (to do forward reasoning).
+   Specifying multiple :n:`@one_term {? with @bindings }` is equivalent to
+   giving each one serially, left to right, as separate `apply` tactics.
+
+   The type of :n:`@one_term` consists of zero or more :term:`premises <premise>`
+   followed by a :ref:`conclusion <conclusion_meaning_2>`,
+   i.e. it has the form :n:`{* @term__premise -> } @term__conclusion`.
+   Backward reasoning with a :n:`@one_term` whose type is, for example, `A -> B`
+   replaces an as-yet unproven goal `B` with `A`.  Forward reasoning with the same
+   :n:`@one_term` changes a hypothesis with type `A` to `B`.  (Hypotheses are
+   proven propositions that are valid only within the context that contains them.)
+
+   Unification creates a map from the variables in the type of :n:`@one_term`
+   to matching subterms of the goal or hypothesis.
+   The matching subterms are then substituted into the type of :n:`@one_term`
+   when generating the updated goal or hypothesis.  Unmatched premises become
+   new subgoals with similar substitutions.  If no match is found, the
+   tactic fails.
+
+   The goal and hypothesis cases are described separately for clarity.
+
+   :n:`@one_term`
+     If :n:`@one_term` is an :n:`@ident`, it is the name of
+     a theorem, lemma or hypothesis whose type is given in the
+     theorem statement or shown in the context.  Otherwise it is a proof term whose
+     type can be displayed with :cmd:`Check` :n:`@one_term`.
+
+   Without :n:`@in_hyp_as` (the goal case)
+     If the goal matches all of the type of :n:`@one_term` (both premises and
+     the conclusion), the tactic proves the goal.
+     Otherwise, the tactic matches the goal against the conclusion of :n:`@one_term`.
+     If the match succeeds, the tactic replaces the current goal with a subgoal for
+     each non-dependent premise of the type of :n:`@one_term`.  See the example
+     :ref:`here <apply_backward>`.
+
+     If the conclusion of the type of :token:`one_term` does
+     not match the goal *and* the conclusion is an inductive type isomorphic to
+     a tuple type, then each component of the tuple is recursively matched to
+     the goal in the left-to-right order.
+
+     .. todo PR: does the next paragraph apply to the hypothesis case, too?
+        Also, can one of you help me create an example?
```

Comment:
I added the first example.  Not sure the second example adds much for the reader and the first example alone is a bit lengthy.

Reply:
Is second-order unification supported primarily as a way to get around issues like this?

Reply:
I'm not completely sure what counts as second-order unification, whether the first example is already considered as such or only the second example where the unification has created the `fun ...` term that it instantiated `P` with. Second-order unification is just about making the call to `apply nat_ind` succeed. But it is also not very smart as highlighted in these two examples where the default instantiation is not the expected one. I don't have an example where `apply nat_ind` would choose the expected instantiation (again, these are somehow artificial examples since usually users should use `induction` which is smarter).

Reply:
"The goal case uses first-order unification with dependent types unless ..." pretty clearly suggests that the special case is not first-order unification.  I mention second-order unification in the example I added, if you think we should avoid the term, I can remove that after you look at it.

Reply:
We can consider this to be a case of second-order unification indeed. According to https://en.wikipedia.org/wiki/Unification_(computer_science), unification is higher order as soon as some variables are matched to functions. The only drawback of not putting the second example is that it doesn't highlight that variables can be unified to a new `fun ...` pattern created on the spot.

Reply:
Hmm, even looking closely at the successful `apply` tactics and `Show Proof` steps in both examples, I don't quite see what you mean by "variables can be unified to a new `fun ...` pattern created on the spot", so it may be hard to explain the difference between the two examples.  I'll just use the first example.

Reply:
This is what happens in my second example before the use of `pattern`: `(fun _ : nat => x + y = y + x /\ True)` did not exist before in the goal. But fine to skip this.